### ISSUE TITLE
SF 4.2 deprecation notices re: ContainerAwareCommand

### ISSUE THREAD

**Issue by hal869** [FriendsOfSymfony/FOSHttpCacheBundle#505]
EDIT: Removing ContainerAware would be a BC break for older Symfony versions, and requiring Constructor injection is a BC break for everybody configuring their own services. The code is ready to remove ContainerAware, lets remove it when we bump the bundle to version 3.

Since upgrade to Symfony 4.2, seeing these Deprecation notices,
```
The "Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand" class is deprecated since Symfony 4.2, use "Symfony\Component\Console\Command\Command" with dependency injection instead.

[▼
  "exception" => SilencedErrorContext {#4540 ▼
    +count: 1
    -severity: E_USER_DEPRECATED
    trace: {▼
      /srv/web/test/vendor/symfony/framework-bundle/Command/ContainerAwareCommand.php:18 {▼
        › 
        › @trigger_error(sprintf('The "%s" class is deprecated since Symfony 4.2, use "%s" with dependency injection instead.', ContainerAwareCommand::class, Command::class), E_USER_DEPRECATED);
        › 
      }
      /srv/web/test/vendor/friendsofsymfony/http-cache-bundle/src/Command/BaseInvalidateCommand.php:22 {▼
        ›  */
        › abstract class BaseInvalidateCommand extends ContainerAwareCommand
        › {
      }
    }
  }
]

{▼
  /srv/web/test/vendor/symfony/framework-bundle/Command/ContainerAwareCommand.php:18 {▼
    › 
    › @trigger_error(sprintf('The "%s" class is deprecated since Symfony 4.2, use "%s" with dependency injection instead.', ContainerAwareCommand::class, Command::class), E_USER_DEPRECATED);
    › 
  }
  /srv/web/test/vendor/friendsofsymfony/http-cache-bundle/src/Command/BaseInvalidateCommand.php:22 {▼
    ›  */
    › abstract class BaseInvalidateCommand extends ContainerAwareCommand
    › {
  }
}
```

and
```
The "FOS\HttpCacheBundle\Command\BaseInvalidateCommand" class extends "Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand" that is deprecated since Symfony 4.2, use {@see Command} instead.

[▼
  "exception" => SilencedErrorContext {#4539 ▼
    +count: 1
    -severity: E_USER_DEPRECATED
    trace: {▼
      /srv/web/test/vendor/symfony/debug/DebugClassLoader.php:191 {▶}
      /srv/web/test/vendor/friendsofsymfony/http-cache-bundle/src/Command/InvalidatePathCommand.php:24 {▼
        ›  */
        › class InvalidatePathCommand extends BaseInvalidateCommand
        › {
      }
    }
  }
]

{▼
  /srv/web/test/vendor/symfony/debug/DebugClassLoader.php:191 {▼
    › foreach ($deprecations as $message) {
    ›     @trigger_error($message, E_USER_DEPRECATED);
    › }
  }
  /srv/web/test/vendor/friendsofsymfony/http-cache-bundle/src/Command/InvalidatePathCommand.php:24 {▼
    ›  */
    › class InvalidatePathCommand extends BaseInvalidateCommand
    › {
  }
}
```

Here's the relevant deprecation notice:

https://symfony.com/blog/new-in-symfony-4-2-important-deprecations#deprecated-containerawarecommand

**Comment by Toflar** [comment]
Working on a PR https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/pull/508

**Comment by Toflar** [comment]
This can be closed then. Do you mind releasing 2.6.1? :)

**Comment by dbu** [comment]
yep, willdo

**Comment by tjamps** [comment]
I think there is a misunderstanding here : PR #508 fixes deprecation warnings about `TreeBuilder` requiring a rootNode when instatiated ; yet @hal869 opened this issue regarding `The "Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand" class is deprecated since Symfony 4.2, use "Symfony\Component\Console\Command\Command" with dependency injection instead.`

I can confirm this still happens today with : 

- `friendsofsymfony/http-cache-bundle  2.6.1`
- `symfony/framework-bundle v4.2.5`

Is it possible to re-open the issue so that we can have a look at this ?

**Comment by dbu** [comment]
oh, you are right. happy if you can do a pull request to adjust this.

**Comment by tjamps** [comment]
Thank you for re-opening the issue :)

I can do a PR in theory. But from my understanding, this would be a BC, meaning the fix would only be available in version 3 of the bundle.

Looking at src/Command/BaseInvalidateCommand.php :

```php
abstract class BaseInvalidateCommand extends ContainerAwareCommand
{
    private $cacheManager;

    public function __construct(CacheManager $cacheManager = null)
    {
        if (!$cacheManager) {
            @trigger_error('Instantiating commands without the cache manager is deprecated and will be removed in version 3', E_USER_DEPRECATED);
        }
        $this->cacheManager = $cacheManager;
        parent::__construct();
    }

    protected function getCacheManager()
    {
        if (!$this->cacheManager) {
            $this->cacheManager = $this->getContainer()->get('fos_http_cache.cache_manager');
        }

        return $this->cacheManager;
    }
}
```

What do you think ?

**Comment by dbu** [comment]
right. so we are basically ready for this, but need to do a BC break for it. imho we don't have to hurry with a version 3 because of this, its only a deprecation, not a bug.

version 2 is almost 2 years old, july 2017 - but i'd prefer to aim for version 3 when symfony 5 comes out, so we can do other BC breaks if necessary. i expect that to be end of 2019.

i'll start a version 3 tag and add this issue to it.

**Comment by tjamps** [comment]
Sounds good to me, thanks for the feedback.

I'll be more than happy to join the effort to support Symfony 5 :)

**Comment by dbu** [comment]
@XWB not sure what you mean. we know that there is a deprecation warning and as discussed in the thread above your post, can't be fixed without a BC break in the bundle. so we wait with fixing until we release a version of the bundle that breaks BC and jumps to recent symfony versions only. is there a problem with that approach in your opinion?

**Comment by followyounger** [comment]
[2019-08-17 08:17:02] php.INFO: User Deprecated: The "Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand" class is deprecated since Symfony 4.2, use "Symfony\Component\Console\Command\Command" with dependency injection instead. {"exception":"[object] (ErrorException(code: 0): User Deprecated: The \"Symfony\\Bundle\\FrameworkBundle\\Command\\ContainerAwareCommand\" class is deprecated since Symfony 4.2, use \"Symfony\\Component\\Console\\Command\\Command\" with dependency injection instead. at /data/home/www/operations/src/v20190816193531/vendor/symfony/framework-bundle/Command/ContainerAwareCommand.php:18)"} []
[2019-08-17 08:17:11] php.INFO: User Deprecated: The "excluded_404s" option is deprecated in MonologBundle since version 3.4.0, you should rely on the "excluded_http_codes" option instead. {"exception":"[object] (ErrorException(code: 0): User Deprecated: The \"excluded_404s\" option is deprecated in MonologBundle since version 3.4.0, you should rely on the \"excluded_http_codes\" option instead. at /data/home/www/operations/src/v20190816193531/vendor/symfony/monolog-bundle/DependencyInjection/MonologExtension.php:381)"} []
[2019-08-17 08:17:11] php.INFO: User Deprecated: A tree builder without a root node is deprecated since Symfony 4.2 and will not be supported anymore in 5.0. {"exception":"[object] (ErrorException(code: 0): User Deprecated: A tree builder without a root node is deprecated since Symfony 4.2 and will not be supported anymore in 5.0. at /data/home/www/operations/src/v20190816193531/vendor/symfony/config/Definition/Builder/TreeBuilder.php:30)"} []
[2019-08-17 08:17:11] php.INFO: User Deprecated: A tree builder without a root node is deprecated since Symfony 4.2 and will not be supported anymore in 5.0. {"exception":"[object] (ErrorException(code: 0): User Deprecated: A tree builder without a root node is deprecated since Symfony 4.2 and will not be supported anymore in 5.0. at /data/home/www/operations/src/v20190816193531/vendor/symfony/config/Definition/Builder/TreeBuilder.php:30)"} []
[2019-08-17 08:17:11] php.INFO: User Deprecated: A tree builder without a root node is deprecated since Symfony 4.2 and will not be supported anymore in 5.0. {"exception":"[object] (ErrorException(code: 0): User Deprecated: A tree builder without a root node is deprecated since Symfony 4.2 and will not be supported anymore in 5.0. at /data/home/www/operations/src/v20190816193531/vendor/symfony/config/Definition/Builder/TreeBuilder.php:30)"} []
[2019-08-17 08:17:11] php.INFO: User Deprecated: A tree builder without a root node is deprecated since Symfony 4.2 and will not be supported anymore in 5.0. {"exception":"[object] (ErrorException(code: 0): User Deprecated: A tree builder without a root node is deprecated since Symfony 4.2 and will not be supported anymore in 5.0. at /data/home/www/operations/src/v20190816193531/vendor/symfony/config/Definition/Builder/TreeBuilder.php:30)"} []
[2019-08-17 08:17:12] php.INFO: User Deprecated: The "Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand" class is deprecated since Symfony 4.2, use "Symfony\Component\Console\Command\Command" with dependency injection instead. {"exception":"[object] (ErrorException(code: 0): User Deprecated: The \"Symfony\\Bundle\\FrameworkBundle\\Command\\ContainerAwareCommand\" class is deprecated since Symfony 4.2, use \"Symfony\\Component\\Console\\Command\\Command\" with dependency injection instead. at /data/home/www/operations/src/v20190816193531/vendor/symfony/framework-bundle/Command/ContainerAwareCommand.php:18)"} []

**Comment by dbu** [comment]
@followyounger the configuration class has been updated a while ago: https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/blob/master/src/DependencyInjection/Configuration.php#L59-L63

are you on the latest version of the bundle? if so, are you sure that this warning is not from some other bundle? you are welcome to make pull requests to improve things, or track down specific things, but pasting this is not very helpful.

**Comment by dbu** [comment]
fixed in #529
